### RFM 분석 실습
- Recency: 얼마나 최근에 구매했는가
- Frequency: 얼마나 자주 구매했는가
- Monetary: 얼마나 많은 금액을 지출했는가

In [1]:
# 빈도수가 부족하다면 구매한수량으로 대체

### 고객 분석
- CustomerID: 고객 번호
- PurchaseDate: 구매 날짜
- TransactionAmount: 결제 금액 
- ProductInformation: 상품 정보
- OrderID: 주문 번호
- Location: 주문 지역

In [10]:
import pandas as pd
cp_df = pd.read_csv('./datasets/customer_order_data.csv')
display(cp_df)

CustomerID PurchaseDate  TransactionAmount ProductInformation  OrderID  \
0          8814   2023-04-11             943.31          Product C   890075   
1          2188   2023-04-11             463.70          Product A   176819   
2          4608   2023-04-11              80.28          Product A   340062   
3          2559   2023-04-11             221.29          Product A   239145   
4          9482   2023-04-11             739.56          Product A   194545   
..          ...          ...                ...                ...      ...   
995        2970   2023-06-10             759.62          Product B   275284   
996        6669   2023-06-10             941.50          Product C   987025   
997        8836   2023-06-10             545.36          Product C   512842   
998        1440   2023-06-10             729.94          Product B   559753   
999        4759   2023-06-10             804.28          Product D   467544   

     Location  
0       Tokyo  
1      London  
2    New York  
3      London  
4       Paris  
..        ...  
995    London  
996  New York  
997    London  
998     Paris  
999  New York  

[1000 rows x 6 columns]

In [11]:
display(cp_df.head(10)) # 상위 10개 추출 

CustomerID PurchaseDate  TransactionAmount ProductInformation  OrderID  \
0        8814   2023-04-11             943.31          Product C   890075   
1        2188   2023-04-11             463.70          Product A   176819   
2        4608   2023-04-11              80.28          Product A   340062   
3        2559   2023-04-11             221.29          Product A   239145   
4        9482   2023-04-11             739.56          Product A   194545   
5        8483   2023-04-11             375.23          Product C   691194   
6        8317   2023-04-11             272.56          Product B   826847   
7        6911   2023-04-11             433.33          Product C   963918   
8        8993   2023-04-12              16.55          Product D   112426   
9        3519   2023-04-12             464.63          Product C   139726   

   Location  
0     Tokyo  
1    London  
2  New York  
3    London  
4     Paris  
5     Paris  
6  New York  
7     Tokyo  
8  New York  
9  New York

In [4]:
display(cp_df.tail(10)) # 하위 10개 추출 

CustomerID PurchaseDate  TransactionAmount ProductInformation  OrderID  \
990        6570   2023-06-10             125.84          Product C   664032   
991        1029   2023-06-10             704.99          Product D   992083   
992        5938   2023-06-10             219.12          Product B   901280   
993        1990   2023-06-10             791.40          Product C   759104   
994        4982   2023-06-10             624.79          Product B   972881   
995        2970   2023-06-10             759.62          Product B   275284   
996        6669   2023-06-10             941.50          Product C   987025   
997        8836   2023-06-10             545.36          Product C   512842   
998        1440   2023-06-10             729.94          Product B   559753   
999        4759   2023-06-10             804.28          Product D   467544   

     Location  
990     Paris  
991  New York  
992     Paris  
993     Tokyo  
994     Paris  
995    London  
996  New York  
997    London  
998     Paris  
999  New York

In [12]:
cp_df.info() # cp DataFrame 정보 확인 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   CustomerID          1000 non-null   int64  
 1   PurchaseDate        1000 non-null   object 
 2   TransactionAmount   1000 non-null   float64
 3   ProductInformation  1000 non-null   object 
 4   OrderID             1000 non-null   int64  
 5   Location            1000 non-null   object 
dtypes: float64(1), int64(2), object(3)
memory usage: 47.0+ KB


In [13]:
# 중복행 검사
cp_df.duplicated().sum() 

0

In [14]:
# 결측치가 모두 0 인 것 을 확인 
cp_df.isnull().sum()
cp_df.isna

<bound method DataFrame.isna of      CustomerID PurchaseDate  TransactionAmount ProductInformation  OrderID  \
0          8814   2023-04-11             943.31          Product C   890075   
1          2188   2023-04-11             463.70          Product A   176819   
2          4608   2023-04-11              80.28          Product A   340062   
3          2559   2023-04-11             221.29          Product A   239145   
4          9482   2023-04-11             739.56          Product A   194545   
..          ...          ...                ...                ...      ...   
995        2970   2023-06-10             759.62          Product B   275284   
996        6669   2023-06-10             941.50          Product C   987025   
997        8836   2023-06-10             545.36          Product C   512842   
998        1440   2023-06-10             729.94          Product B   559753   
999        4759   2023-06-10             804.28          Product D   467544   

     Location  
0  

In [15]:
display(cp_df.describe().T) # 데이터 수와 평균,표준편차, 최소,최대값, 사분위 분포도 추출

count          mean            std        min        25%  \
CustomerID         1000.0    5554.78900    2605.014863    1011.00    3273.00   
TransactionAmount  1000.0     513.67781     286.098700      12.13     257.12   
OrderID            1000.0  554071.39800  264695.448814  100096.00  313152.00   

                          50%        75%        max  
CustomerID           5538.000    7821.75    9991.00  
TransactionAmount     523.565     759.86     999.44  
OrderID            564671.500  783052.25  999695.00

In [18]:
# 무엇을 기준점으로 삼을 것인가 -> 결제 금액 
import numpy as np

Q1 = cp_df.describe().T.loc['TransactionAmount', '25%']
Q3 = cp_df.describe().T.loc['TransactionAmount', '75%']

iqr = Q3 - Q1

lower_bound = Q1 - iqr * 1.5
upper_bound = Q3 + iqr * 1.5

if lower_bound < 0:
    lower_bound = 0

print(f'정상치 범위: {lower_bound} ~ {upper_bound}')


정상치 범위: 0 ~ 1513.97


In [19]:
# 이상치를 제거 할 것인가 ? -> 일단 해보자 --> 이상치에 해당하는 값이 없다
delete_upper_df = cp_df[cp_df['TransactionAmount'] > upper_bound]
delete_upper_df

delete_lower_df = cp_df[cp_df['TransactionAmount'] < lower_bound]
delete_lower_df

Empty DataFrame
Columns: [CustomerID, PurchaseDate, TransactionAmount, ProductInformation, OrderID, Location]
Index: []

In [20]:
# feature명 변경 
cp_df = cp_df.rename(columns = {'ProductInformation':'Product'})
cp_df = cp_df.rename(columns = {'TransactionAmount':'Transaction'})
cp_df = cp_df.rename(columns = {'PurchaseDate' : 'date'})
cp_df = cp_df.rename(columns = {'OrderID' : 'Order'})
cp_df = cp_df.rename(columns = {'CustomerID' : 'Customer'})
cp_df

Customer        date  Transaction    Product   Order  Location
0        8814  2023-04-11       943.31  Product C  890075     Tokyo
1        2188  2023-04-11       463.70  Product A  176819    London
2        4608  2023-04-11        80.28  Product A  340062  New York
3        2559  2023-04-11       221.29  Product A  239145    London
4        9482  2023-04-11       739.56  Product A  194545     Paris
..        ...         ...          ...        ...     ...       ...
995      2970  2023-06-10       759.62  Product B  275284    London
996      6669  2023-06-10       941.50  Product C  987025  New York
997      8836  2023-06-10       545.36  Product C  512842    London
998      1440  2023-06-10       729.94  Product B  559753     Paris
999      4759  2023-06-10       804.28  Product D  467544  New York

[1000 rows x 6 columns]

In [21]:
display(cp_df)

Customer        date  Transaction    Product   Order  Location
0        8814  2023-04-11       943.31  Product C  890075     Tokyo
1        2188  2023-04-11       463.70  Product A  176819    London
2        4608  2023-04-11        80.28  Product A  340062  New York
3        2559  2023-04-11       221.29  Product A  239145    London
4        9482  2023-04-11       739.56  Product A  194545     Paris
..        ...         ...          ...        ...     ...       ...
995      2970  2023-06-10       759.62  Product B  275284    London
996      6669  2023-06-10       941.50  Product C  987025  New York
997      8836  2023-06-10       545.36  Product C  512842    London
998      1440  2023-06-10       729.94  Product B  559753     Paris
999      4759  2023-06-10       804.28  Product D  467544  New York

[1000 rows x 6 columns]

In [22]:
cp_df['Product'].value_counts()

Product
Product C    277
Product D    253
Product B    245
Product A    225
Name: count, dtype: int64

In [23]:
# 상품정보 value명을 변경 A,B,C,D
cp_df['Product'] = cp_df['Product'].replace(
    ['Product A', 'Product B', 'Product C', 'Product D'], ['A', 'B', 'C', 'D']
)

cp_df['Product'].value_counts()

Product
C    277
D    253
B    245
A    225
Name: count, dtype: int64

In [24]:
display(cp_df)

Customer        date  Transaction Product   Order  Location
0        8814  2023-04-11       943.31       C  890075     Tokyo
1        2188  2023-04-11       463.70       A  176819    London
2        4608  2023-04-11        80.28       A  340062  New York
3        2559  2023-04-11       221.29       A  239145    London
4        9482  2023-04-11       739.56       A  194545     Paris
..        ...         ...          ...     ...     ...       ...
995      2970  2023-06-10       759.62       B  275284    London
996      6669  2023-06-10       941.50       C  987025  New York
997      8836  2023-06-10       545.36       C  512842    London
998      1440  2023-06-10       729.94       B  559753     Paris
999      4759  2023-06-10       804.28       D  467544  New York

[1000 rows x 6 columns]

In [25]:
# 상품정보 별로 금액을 묶어보자
# 상품 정보 별 평균 금액 
cp_df_a = cp_df.groupby('Product')['Transaction'].mean().reset_index()
display(cp_df_a)

# 상품 정보 별 최대 금액 
cp_df_b = cp_df.groupby('Product')['Transaction'].max().reset_index()
display(cp_df_b)

# 상품 정보 별 최소 금액 
cp_df_c = cp_df.groupby('Product')['Transaction'].min().reset_index()
display(cp_df_c)

# 상품 정보 별 금액 표준편차
cp_df_d = cp_df.groupby('Product')['Transaction'].std().reset_index()
display(cp_df_d)


Product  Transaction
0       A   521.866356
1       B   502.196245
2       C   521.629928
3       D   508.807549

Product  Transaction
0       A       999.29
1       B       997.19
2       C       999.44
3       D       996.98

Product  Transaction
0       A        12.71
1       B        13.51
2       C        17.15
3       D        12.13

Product  Transaction
0       A   291.742834
1       B   273.062067
2       C   293.223626
3       D   286.677402

#### 1차 분석 (상품정보 & 상품금액)
- 전반적으로 분석을 해보았을때, 평균적으로 가장 금액이 높은 상품은 A그룹이며, 그 다음으로 C, D, B 그룹순이다.
- 상품별 최대 금액은 C 그룹에서 999.44로 가장 높게 나왔으며 최소 금액은 D 그룹에서 12.13으로 가장 적은 최소값을 기록하였다.
- 그러나 한가지 의문점은 평균적으로 가장 가격이 높게 책정이 되었던 A 그룹에서 최소값 측정 결과 12.71로 하위 2번째로 최소값이
  낮았으며, 이를 미루어보았을 때, A그룹의 표준편차가 가장 클 것으로 예상이 되었다.
- 하지만 평균가격이 2순위였으며, 최대금액이 1순위 였던 C그룹의 표준편차가 가장 큰 것을 알 수 있었고, 아마 상품정보 말고도 가격과 연관된 정보가 있을 것으로 판단하여, 고객번호도 같이 연관을 지어서 분석을 하기로 하였다.


In [26]:
# 고객번호의 빈도수를 Frequency로 엮어보자!
# 몇가지 문제점 -> 고객은 굉장히 데이터가 많을 텐데 이 고객정보를 어떻게 그룹으로 묶어서 Montery로 
# 만들 수 있을까 라는 의문점이 생겼다. -> 고객정보와 상품정보를 같이 frequency_array에 담아서 집계함수로 총합을 구해보자.
# 그 전에 우선 고객정보의 수는 매우 많고, 중복된 부분도 많기 때문에 우선 중복된 고객정보를 지우고 다 하나로 초기화하자
cp_df['Order'].value_counts()

Order
890075    1
790658    1
778720    1
296412    1
621841    1
         ..
540218    1
651231    1
225245    1
291663    1
467544    1
Name: count, Length: 1000, dtype: int64

In [27]:
# 고객 정보는 중복된 행이 있기때문에 이를 제거한다.
cp_df = cp_df.drop_duplicates(subset=['Customer'])
cp_df['Customer'].value_counts()

Customer
8814    1
9761    1
3719    1
9383    1
7457    1
       ..
8625    1
6707    1
2075    1
5438    1
4759    1
Name: count, Length: 946, dtype: int64

In [28]:
# grouped -> 지역별로 멤버를 담아놓은 변량 
grouped = cp_df.groupby('Location')['Customer'].apply(list).reset_index()
print(grouped)

# producted -> 상품별로 멤버를 담아놓은 변량 
producted = cp_df.groupby('Product')['Customer'].apply(list).reset_index()
print(producted)

# trancationed -> 고객별로 구매한 결제 금액 
trancationed = cp_df.groupby('Customer')['Transaction'].apply(list).reset_index()
print(trancationed)

# 상품 결제 최고 금액 (회원 아이디:9991, 결제 금액 : 999.44)
max_price = trancationed.sum()
print(max_price)

display(cp_df)

   Location                                           Customer
0    London  [2188, 2559, 9409, 5670, 8724, 2824, 8705, 634...
1  New York  [4608, 8317, 8993, 3519, 5837, 1520, 6878, 958...
2     Paris  [9482, 8483, 9005, 9128, 8719, 5533, 9871, 986...
3     Tokyo  [8814, 6911, 9148, 2767, 8529, 2228, 9115, 558...
  Product                                           Customer
0       A  [2188, 4608, 2559, 9482, 9409, 8705, 5173, 430...
1       B  [8317, 5670, 2824, 1520, 8719, 8529, 2228, 553...
2       C  [8814, 8483, 6911, 3519, 9005, 8724, 4161, 663...
3       D  [8993, 9148, 5837, 2767, 9128, 6347, 1146, 261...
     Customer Transaction
0        1011    [918.72]
1        1025    [359.29]
2        1029    [704.99]
3        1046    [859.82]
4        1049    [225.72]
..        ...         ...
941      9941    [960.53]
942      9950    [679.11]
943      9954    [798.01]
944      9985      [36.1]
945      9991    [626.81]

[946 rows x 2 columns]
Customer                                    

Customer        date  Transaction Product   Order  Location
0        8814  2023-04-11       943.31       C  890075     Tokyo
1        2188  2023-04-11       463.70       A  176819    London
2        4608  2023-04-11        80.28       A  340062  New York
3        2559  2023-04-11       221.29       A  239145    London
4        9482  2023-04-11       739.56       A  194545     Paris
..        ...         ...          ...     ...     ...       ...
995      2970  2023-06-10       759.62       B  275284    London
996      6669  2023-06-10       941.50       C  987025  New York
997      8836  2023-06-10       545.36       C  512842    London
998      1440  2023-06-10       729.94       B  559753     Paris
999      4759  2023-06-10       804.28       D  467544  New York

[946 rows x 6 columns]

In [32]:

# 각 고객의 총 구매 횟수 계산
cp_df['Monetary'] = cp_df.groupby('Customer')['Transaction'].sum().reset_index()
cp_df
## Montery 추출 완성 

ValueError: Columns must be same length as key

In [ ]:
## Montery 이상치 조사 
import numpy as np

Q1 = cp_df.describe().T.loc['Monetary', '25%']
Q3 = cp_df.describe().T.loc['Monetary', '75%']

iqr = Q3 - Q1

lower_bound = Q1 - iqr * 1.5
uppper_bound = Q3 + iqr * 1.5

if lower_bound < 0:
    lower_bound = 0

print(f'정상치 범위: {lower_bound} ~ {uppper_bound}')



In [31]:
mean_df_location = cp_df.groupby('Location')['Monetary'].mean().reset_index()
max_df_location = cp_df.groupby('Location')['Monetary'].max().reset_index()
min_df_location = cp_df.groupby('Location')['Monetary'].min().reset_index()
std_df_location = cp_df.groupby('Location')['Monetary'].std().reset_index()
display(mean_df_location)
display(max_df_location)
display(min_df_location)
display(std_df_location)

Location  Monetary
0    London       NaN
1  New York       NaN
2     Paris       NaN
3     Tokyo       NaN

Location  Monetary
0    London       NaN
1  New York       NaN
2     Paris       NaN
3     Tokyo       NaN

Location  Monetary
0    London       NaN
1  New York       NaN
2     Paris       NaN
3     Tokyo       NaN

Location  Monetary
0    London       NaN
1  New York       NaN
2     Paris       NaN
3     Tokyo       NaN

In [30]:
display(cp_df)

Customer        date  Transaction Product   Order  Location  Monetary
0        8814  2023-04-11       943.31       C  890075     Tokyo       NaN
1        2188  2023-04-11       463.70       A  176819    London       NaN
2        4608  2023-04-11        80.28       A  340062  New York       NaN
3        2559  2023-04-11       221.29       A  239145    London       NaN
4        9482  2023-04-11       739.56       A  194545     Paris       NaN
..        ...         ...          ...     ...     ...       ...       ...
995      2970  2023-06-10       759.62       B  275284    London       NaN
996      6669  2023-06-10       941.50       C  987025  New York       NaN
997      8836  2023-06-10       545.36       C  512842    London       NaN
998      1440  2023-06-10       729.94       B  559753     Paris       NaN
999      4759  2023-06-10       804.28       D  467544  New York       NaN

[946 rows x 7 columns]